### PR TITLE
ENH: Implement DataFrame.value_counts

### LINKED ISSUE #5377: ENH:  DataFrame.value_counts()
Enable value_counts on DataFrame by delegating along axis (ie columns/items) to value_counts as convenience, boiling down to same Series value_counts (Panel is more complicated so not going to attempt for now).

Gets weird with heterogeneous dtypes (b/c union of all unique values) 

@rockg - can you make up some test cases? I already have the implementation. I can put together panel.

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] closes #5377
- [x] tests added / passed
- [x] passes `black pandas`
- [x] whatsnew entry

This is picking up where https://github.com/pandas-dev/pandas/pull/27350 left off because I think it'd be a nice feature to have. At least one thing that still needs to be done is implementing `bins` when we have only a single column in `subset`, in which case maybe we can just delegate to `Series.value_counts`.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -5064,6 +5064,120 @@ def sort_index(
         else:
             return self._constructor(new_data).__finalize__(self)
 
+    def value_counts(
+        self,
+        subset=None,
```
Comment (WillAyd) [inline pandas/core/frame.py]:
```suggestion
        subset: Sequence[Label] = None,
```

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5064,6 +5064,120 @@ def sort_index(
...
+        normalize: bool = False,
+        sort: bool = True,
+        ascending: bool = False,
+        bins=None,
```
Comment (WillAyd) [inline pandas/core/frame.py]:
```suggestion
        bins: Optional[int] = None,
```

**Review by WillAyd — CHANGES_REQUESTED** [review]
some minor nits on annotations but otherwise lgtm

File: doc/source/whatsnew/v1.1.0.rst [hunk 3]
```diff
@@ -54,6 +54,7 @@ Other API changes
 
 - :meth:`Series.describe` will now show distribution percentiles for ``datetime`` dtypes, statistics ``first`` and ``last``
   will now be ``min`` and ``max`` to match with numeric dtypes in :meth:`DataFrame.describe` (:issue:`30164`)
+- Added :meth:`DataFrame.value_counts` (:issue:`5377`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: add a doc-link to the section in the docs where you showcase this (need to add)

File: pandas/core/frame.py [hunk 4]
```diff
@@ -5070,6 +5070,120 @@ def sort_index(
...
+        subset: Optional[Sequence[Label]] = None,
+        normalize: bool = False,
+        sort: bool = True,
+        ascending: bool = False,
+        bins: Optional[int] = None,
+        dropna: bool = True,
+    ):
+        """
+        Return a Series containing counts of unique rows in the DataFrame.
+        .. versionadded:: 1.1.0
```
Comment (jreback) [inline pandas/core/frame.py]: i think you need a blank line before/after to render

File: pandas/core/frame.py [hunk 5]
```diff
@@ -5070,6 +5070,120 @@ def sort_index(
...
+        first element is the most frequently-occurring row.
+
+        Parameters
+        ----------
+        subset : list-like, optional
+            Columns to use when counting unique combinations.
+        normalize : bool, default False
+            Return proportions rather than frequencies.
+        sort : bool, default True
+            Sort by frequencies.
```
Comment (jreback) [inline pandas/core/frame.py]: does this argument order follow Series.value_counts?

File: pandas/core/frame.py [hunk 6]
```diff
@@ -5070,6 +5070,120 @@ def sort_index(
...
+        >>> df = pd.DataFrame({'num_legs': [2, 4, 4, 6],
+        ...                    'num_wings': [2, 0, 0, 0]},
+        ...                   index=['falcon', 'dog', 'cat', 'ant'])
+        >>> df
+                num_legs  num_wings
+        falcon         2          2
+        dog            4          0
+        cat            4          0
+        ant            6          0
+        >>> df.value_counts()
```
Comment (jreback) [inline pandas/core/frame.py]: blank line between the cases. Add a comment if appropriate on the cases.

File: pandas/core/frame.py [hunk 7]
```diff
@@ -5070,6 +5070,120 @@ def sort_index(
...
+        dtype: int64
+        >>> df.value_counts(normalize=True)
+        num_legs  num_wings
+        4         0            0.50
+        6         0            0.25
+        2         2            0.25
+        dtype: float64
+        """
+        if subset is None:
+            subset = self.columns.tolist()
```
Comment (jreback) [inline pandas/core/frame.py]: don't need the .tolist()

File: pandas/core/frame.py [hunk 8]
```diff
@@ -5070,6 +5070,120 @@ def sort_index(
...
+
+        # Some features not supported yet
```
Comment (jreback) [inline pandas/core/frame.py]: i would remove these args then

File: pandas/core/frame.py [hunk 9]
```diff
@@ -5070,6 +5070,120 @@ def sort_index(
...
+                "`bins` parameter not yet supported for DataFrames."
+            )
+
+        counts = self.groupby(subset).size()
+
+        if sort:
+            counts = counts.sort_values(ascending=ascending)
+        if normalize:
+            counts /= counts.sum()
+        # Force MultiIndex for single column
```
Comment (jreback) [inline pandas/core/frame.py]: blank line

File: doc/source/whatsnew/v1.1.0.rst (hunk 3 above)
Comment (jreback) [reply doc/source/whatsnew/v1.1.0.rst]: add docs near where we talk about Series.value_counts

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 5 above)
Comment (dsaxton) [reply pandas/core/frame.py]: Yes

File: doc/source/whatsnew/v1.1.0.rst (hunk 3 above)
Comment (dsaxton) [reply doc/source/whatsnew/v1.1.0.rst]: A bit confused by this one. Where should I showcase the method and how to link to it?

File: pandas/core/frame.py (hunk 7 above)
Comment (dsaxton) [reply pandas/core/frame.py]: I think we might need this. Kind of surprising but `df.groupby(df.columns)` raises `ValueError: Grouper and axis must be same length`.

File: pandas/core/frame.py [hunk 10]
```diff
@@ -5079,6 +5079,102 @@ def sort_index(
...
+        normalize: bool = False,
+        sort: bool = True,
+        ascending: bool = False,
+    ):
+        """
+        Return a Series containing counts of unique rows in the DataFrame.
+
+        .. versionadded:: 1.1.0
+
+        The returned Series will have a MultiIndex with one level per input
```
Comment (jreback) [inline pandas/core/frame.py]: These should go in Notes

File: pandas/core/frame.py [hunk 11]
```diff
@@ -5079,6 +5079,102 @@ def sort_index(
...
+        ascending : bool, default False
+            Sort in ascending order.
+
+        Returns
+        -------
+        Series
+
+        See Also
+        --------
+        Series.value_counts: Equivalent method on Series.
```
Comment (jreback) [inline pandas/core/frame.py]: I think you have to format this on the next line (the descr)

File: pandas/core/frame.py (hunk 7 above)
Comment (jreback) [reply pandas/core/frame.py]: hmm ok, yeah it likely interpreters this as a same-length indexer and not as separate columns (which is what you want)

**Review by jreback — CHANGES_REQUESTED** [review]
cc @pandas-dev/pandas-core if any comments

**Comment by jreback** [comment]
also pls merge master

File: pandas/core/frame.py (hunk 11 above)
Comment (dsaxton) [reply pandas/core/frame.py]: Hmm, I think it's okay since other methods have it on the same line?

File: doc/source/getting_started/basics.rst [hunk 12]
```diff
@@ -689,6 +689,15 @@ of a 1D array of values. It can also be used as a function on regular arrays:
    s.value_counts()
    pd.value_counts(data)
 
+The :meth:`~DataFrame.value_counts` method can be used to count combinations across multiple columns.
```
Comment (jreback) [inline doc/source/getting_started/basics.rst]: add a versionadded tag 1.1

File: pandas/core/base.py [hunk 13]
```diff
@@ -1196,6 +1196,7 @@ def value_counts(
         --------
         Series.count: Number of non-NA elements in a Series.
         DataFrame.count: Number of non-NA elements in a DataFrame.
+        DataFrame.value_counts: Equivalent method on DataFrames.
```
Comment (jreback) [inline pandas/core/base.py]: can you also add a to DataFrame.value_counts to DataFrame.drop_duplicates

File: pandas/tests/frame/test_value_counts.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
+import numpy as np
+
+import pandas as pd
```
Comment (jreback) [inline pandas/tests/frame/test_value_counts.py]: move to the methods/ subdir

**Review by jreback — CHANGES_REQUESTED** [review]
small comments, ping on green.

**Comment by dsaxton** [comment]
Getting some seemingly unrelated test failures from `test_value_counts_unique_nunique_null` in `pandas/tests/base/test_ops.py` after merging master (I don't _think_ any of this PR touches code tested there since the `Series` method is staying the same). Could this be related to the refactoring from https://github.com/pandas-dev/pandas/pull/32046?

**Comment by dsaxton** [comment]
@jreback Green after the CI fixes, thanks for the review

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @dsaxton 

very nice!

**Pull request merged**